Index: network-manager/data/NetworkManager-wait-online.service.in
===================================================================
--- network-manager.orig/data/NetworkManager-wait-online.service.in
+++ network-manager/data/NetworkManager-wait-online.service.in
@@ -16,7 +16,7 @@ Before=network-online.target
 # time.
 
 Type=oneshot
-ExecStart=@bindir@/nm-online -s -q
+ExecStart=@bindir@/nm-online -s -q --timeout=120
 RemainAfterExit=yes
 
 # Set $NM_ONLINE_TIMEOUT variable for timeout in seconds.
