Index: network-manager/src/core/devices/wifi/nm-device-wifi.c
===================================================================
--- network-manager.orig/src/core/devices/wifi/nm-device-wifi.c
+++ network-manager/src/core/devices/wifi/nm-device-wifi.c
@@ -2311,20 +2311,17 @@ need_new_wpa_psk(NMDeviceWifi *
 
     g_return_val_if_fail(connection, FALSE);
 
-    /* A bad PSK will cause the supplicant to disconnect during the 4-way handshake */
-    if (old_state != NM_SUPPLICANT_INTERFACE_STATE_4WAY_HANDSHAKE)
-        return FALSE;
+    // get wireless security
 
     s_wsec = nm_connection_get_setting_wireless_security(connection);
     if (s_wsec)
         key_mgmt = nm_setting_wireless_security_get_key_mgmt(s_wsec);
 
-    if (g_strcmp0(key_mgmt, "wpa-psk") == 0) {
-/* -4 (locally-generated WLAN_REASON_DISASSOC_DUE_TO_INACTIVITY) usually
-         * means the driver missed beacons from the AP.  This usually happens
-         * due to driver bugs or faulty power-save management.  It doesn't
-         * indicate that the PSK is wrong.
-         */
+	/* A bad PSK will cause the supplicant to disconnect during the 4-way handshake */
+	/* A bad SAE will cause the supplicant to disconnect during the authentication*/
+	if ((g_strcmp0 (key_mgmt, "wpa-psk") == 0 && old_state == NM_SUPPLICANT_INTERFACE_STATE_4WAY_HANDSHAKE) ||
+		(g_strcmp0 (key_mgmt, "sae") == 0 && old_state == NM_SUPPLICANT_INTERFACE_STATE_AUTHENTICATING)) {
+
 #define LOCAL_WLAN_REASON_DISASSOC_DUE_TO_INACTIVITY -4
         if (disconnect_reason == LOCAL_WLAN_REASON_DISASSOC_DUE_TO_INACTIVITY)
             return FALSE;
