Index: network-manager/src/core/dns/nm-dns-manager.c
===================================================================
--- network-manager.orig/src/core/dns/nm-dns-manager.c
+++ network-manager/src/core/dns/nm-dns-manager.c
@@ -1715,6 +1715,10 @@ _mgr_configs_data_construct(NMDnsManager
             has_default_route_explicit || (priority < 0 && has_default_route_auto);
         ip_data->domains.has_default_route =
             ip_data->domains.has_default_route_exclusive || has_default_route_auto;
+        if (ip_data->ip_config_type == NM_DNS_IP_CONFIG_TYPE_VPN
+                && !nm_l3_config_data_get_searches(ip_data->l3cd, ip_data->addr_family, &num)
+                && !nm_l3_config_data_get_domains(ip_data->l3cd, ip_data->addr_family, &num))
+            ip_data->domains.has_default_route = TRUE;
 
         {
             gs_free char *str1 = NULL;
