Index: network-manager/src/core/main.c
===================================================================
--- network-manager.orig/src/core/main.c	2024-12-23 21:22:54.152557255 +0800
+++ network-manager/src/core/main.c	2025-05-30 11:31:56.633567342 +0800
@@ -472,6 +472,10 @@
 
     if (!_dbus_manager_init(config))
         goto done_no_manager;
+    else {
+        if (!nm_dbus_manager_request_name_sync(nm_dbus_manager_get()))
+            goto done_no_manager;
+    }
 
     nm_linux_platform_setup();
 
@@ -508,8 +512,7 @@
     nm_log_dbg(LOGD_CORE, "setting up local loopback");
     nm_platform_link_change_flags(NM_PLATFORM_GET, 1, IFF_UP, TRUE);
 
-    if (!nm_dbus_manager_request_name_sync(nm_dbus_manager_get()))
-        goto done;
+
 
     success = TRUE;
 
